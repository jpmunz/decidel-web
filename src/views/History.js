import React from "react";
import { connect } from "react-redux";
import {
  getCurrentUserName,
  getCurrentRoundHistory,
  getPreviousRoundsHistory
} from "src/state";
import RoundHistory from "./RoundHistory";
import "./History.scss";
import { produce } from "immer";

export class History extends React.Component {
  constructor(props) {
    super(props);

    this.prepareHistory = this.prepareHistory.bind(this);
  }

  prepareHistory(draft) {
    draft.forEach(item => {
      if (item.userName === this.props.currentUserName) {
        item.userName = "You";
      }
    });

    draft.reverse();
  }

  render() {
    const currentRoundHistory = produce(
      this.props.currentRoundHistory,
      this.prepareHistory
    );
    const previousRoundsHistory = produce(
      this.props.previousRoundsHistory,
      this.prepareHistory
    );

    let content;
    if (currentRoundHistory.length || previousRoundsHistory.length) {
      content = (
        <>
          <h4 className="header">How we got here</h4>
          <div className="current-history">
            <RoundHistory history={currentRoundHistory} />
          </div>
          <RoundHistory history={previousRoundsHistory} />
        </>
      );
    }

    return <div className="History">{content}</div>;
  }
}

export default connect((state, props) => {
  return {
    currentRoundHistory: getCurrentRoundHistory(state, props.id),
    previousRoundsHistory: getPreviousRoundsHistory(state, props.id),
    currentUserName: getCurrentUserName(state)
  };
})(History);
