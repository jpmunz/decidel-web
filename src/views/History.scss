@import "mixins";

.History {
  .header {
  }

  .current-history {
    opacity: 0.6;
  }

  .round-header {
    font-size: 1.3rem;
  }

  ul {
    padding-left: 50px;
    margin: 0;
  }
}
