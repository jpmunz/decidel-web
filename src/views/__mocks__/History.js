import React from "react";

export const MockHistory = jest.fn();

MockHistory.mockReturnValue(<div>Mocked History</div>);

const mock = jest.fn().mockImplementation(MockHistory);

export default mock;
