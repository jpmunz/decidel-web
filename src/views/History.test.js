import React from "react";
import { History } from "./History";
import { render } from "@testing-library/react";

const renderComponent = function({
  id = "22",
  currentRoundHistory = [],
  previousRoundsHistory = [],
  currentUserName = "Bob"
}) {
  return render(
    <History
      id={id}
      currentRoundHistory={currentRoundHistory}
      previousRoundsHistory={previousRoundsHistory}
      currentUserName={currentUserName}
    />
  );
};

const renderTest = options =>
  expect(renderComponent(options).asFragment()).toMatchSnapshot();

test("No history available", () => renderTest({}));

test("Current round history only", () =>
  renderTest({
    currentRoundHistory: [
      { userName: "Bob", action: "ADDED", title: "option1" },
      { userName: "Bob", action: "REMOVED", title: "option2" }
    ]
  }));
test("Previous rounds history only", () =>
  renderTest({
    previousRoundsHistory: [
      { userName: "Bob", action: "ADDED", title: "option4" },
      { userName: "Jon", action: "REMOVED", title: "option5" }
    ]
  }));

test("Full history available", () =>
  renderTest({
    currentRoundHistory: [
      { userName: "Bob", action: "ADDED", title: "option1" },
      { userName: "Bob", action: "REMOVED", title: "option2" }
    ],
    previousRoundsHistory: [
      { userName: "Bob", action: "ADDED", title: "option4" },
      { userName: "Jon", action: "REMOVED", title: "option5" }
    ]
  }));
