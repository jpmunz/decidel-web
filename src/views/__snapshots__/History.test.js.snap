// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Current round history only 1`] = `
<DocumentFragment>
  <div
    class="History"
  >
    <h4
      class="header"
    >
      How we got here
    </h4>
    <div
      class="current-history"
    >
      <div
        class="RoundHistory"
      >
        <ul>
          <li
            class="round-item"
          >
            <img
              alt=""
              src="remove.svg"
            />
             You crossed off option2
          </li>
          <li
            class="round-item"
          >
            <img
              alt=""
              src="wrench.svg"
            />
             You threw a wrench by adding option1
          </li>
        </ul>
      </div>
    </div>
    <div
      class="RoundHistory"
    >
      <ul />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Full history available 1`] = `
<DocumentFragment>
  <div
    class="History"
  >
    <h4
      class="header"
    >
      How we got here
    </h4>
    <div
      class="current-history"
    >
      <div
        class="RoundHistory"
      >
        <ul>
          <li
            class="round-item"
          >
            <img
              alt=""
              src="remove.svg"
            />
             You crossed off option2
          </li>
          <li
            class="round-item"
          >
            <img
              alt=""
              src="wrench.svg"
            />
             You threw a wrench by adding option1
          </li>
        </ul>
      </div>
    </div>
    <div
      class="RoundHistory"
    >
      <ul>
        <li
          class="round-item"
        >
          <img
            alt=""
            src="remove.svg"
          />
           Jon crossed off option5
        </li>
        <li
          class="round-item"
        >
          <img
            alt=""
            src="wrench.svg"
          />
           You threw a wrench by adding option4
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;

exports[`No history available 1`] = `
<DocumentFragment>
  <div
    class="History"
  />
</DocumentFragment>
`;

exports[`Previous rounds history only 1`] = `
<DocumentFragment>
  <div
    class="History"
  >
    <h4
      class="header"
    >
      How we got here
    </h4>
    <div
      class="current-history"
    >
      <div
        class="RoundHistory"
      >
        <ul />
      </div>
    </div>
    <div
      class="RoundHistory"
    >
      <ul>
        <li
          class="round-item"
        >
          <img
            alt=""
            src="remove.svg"
          />
           Jon crossed off option5
        </li>
        <li
          class="round-item"
        >
          <img
            alt=""
            src="wrench.svg"
          />
           You threw a wrench by adding option4
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
